src/main/java/org/radarbase/output/worker/KafkaCleanerWorker.kt

```diff
@@ -0,0 +1,103 @@
+package org.radarbase.output.worker
+
+import org.apache.avro.generic.GenericRecord
+import org.radarbase.output.FileStoreFactory
+import org.radarbase.output.accounting.Accountant
+import org.radarbase.output.source.SourceStorage
+import org.radarbase.output.source.TopicFile
+import org.radarbase.output.util.Timer.time
+import org.radarbase.output.worker.RestructureWorker.Companion.extractRecords
+import org.slf4j.LoggerFactory
+import java.io.Closeable
+import java.io.IOException
+import java.nio.file.Path
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+import java.util.concurrent.atomic.AtomicBoolean
+
+class KafkaCleanerWorker(
+        private val fileStoreFactory: FileStoreFactory,
+        private val accountant: Accountant,
+        private val sourceStorage: SourceStorage,
+        private val closed: AtomicBoolean
+) : Closeable {
+    private val maxFilesPerTopic: Int = fileStoreFactory.config.worker.maxFilesPerTopic ?: Int.MAX_VALUE
+    private val pathFactory = fileStoreFactory.pathFactory
+    private val reader = sourceStorage.createReader()
+
+    private val deleteThreshold: Instant? = Instant.now().apply { minus(fileStoreFactory.config.cleaner.age.toLong(), ChronoUnit.DAYS) }
+
+    private val cacheStore = ContainsFileCacheStore(fileStoreFactory)
+
+    fun deleteOldFiles(
+            topic: String,
+            topicPath: Path
+    ): Int {
+        val offsets = accountant.offsets.copyForTopic(topic)
+        return sourceStorage.walker.walkRecords(topic, topicPath)
+                .filter { f -> f.lastModified.isBefore(deleteThreshold) &&
+                        // ensure that there is a file with a larger offset also
+                        // processed, so the largest offset is never removed.
+                        offsets.contains(f.range.copy(range = f.range.range.copy(to = f.range.range.to + 1))) }
+                .take(maxFilesPerTopic)
+                .takeWhile { !closed.get() }
+                .count { file ->
+                    if (isExtracted(file)) {
+                        logger.info("Removing {}", file.path)
+                        time("cleaner.delete") {
+                            sourceStorage.delete(file.path)
+                        }
+                        true
+                    } else {
+                        logger.warn("Source file was not completely extracted: {}", file.path)
+                        accountant.remove(file.range)
+                        false
+                    }
+                }
+    }
+
+    private fun isExtracted(file: TopicFile): Boolean {
```

Comment:
Can we abstract this to another class since the cleaner and data integrity check are different? We can also add a different type of check in the future if needed

Reply:
We may even just want to check the integrity without deleting/cleaning data (for example if we see discrepancies in data and just want to check if everything was extracted successfully. We may intend to keep data still in HDFS for data analysis directly eg using Spark, etc)

Reply:
For that last check, we'd have to keep a separate check-log probably, because otherwise the integrity check will increase with the source data size.

Reply:
it is not needed now, just an idea where separating the two could be useful. But I agree, we can defer it until such a need arises.